pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,606 @@
+package ingressip
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/registry/service/allocator"
+	"k8s.io/kubernetes/pkg/registry/service/ipallocator"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	// It's necessary to allocate for the initial sync in consistent
+	// order rather than in the order received.  This requires waiting
+	// until the initial sync has been processed, and to avoid a hot
+	// loop, we'll wait this long between checks.
+	SyncProcessedPollPeriod = 100 * time.Millisecond
+
+	clientRetryCount    = 5
+	clientRetryInterval = 5 * time.Second
+)
+
+// IngressIPController is responsible for allocating ingress ip
+// addresses to Service objects of type LoadBalancer.
+type IngressIPController struct {
+	client kclient.ServicesNamespacer
+
+	// Services that need to be checked
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	cache      cache.Store
+	controller *framework.Controller
+
+	// Tracks ip allocation for the configured range
+	ipAllocator *ipallocator.Range
+	// Tracks ip -> service key to allow detection of duplicate ip
+	// allocations.
+	allocationMap map[string]string
+
+	// Protects the transition between initial sync and regular processing
+	mu sync.RWMutex
```

Comment:
Done.  

I've only added the queue and cache under the lock, though, since they are the only variables in need of protection (potentially conflicting access between enqueueService and processInitialSync).